File: testware/tck/src/test/scala/org/opencypher/gremlin/tck/GremlinCypherValueConverter.scala
/*
 * Copyright (c) 2018 "Neo4j, Inc." [https://neo4j.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.opencypher.gremlin.tck

import java.{lang, util}

import org.apache.tinkerpop.gremlin.driver.exception.ResponseException
import org.apache.tinkerpop.gremlin.structure.{Edge, Property, Vertex}
import org.opencypher.gremlin.translation.CypherAst
import org.opencypher.tools.tck.api.{CypherValueRecords, ExecutionFailed}
import org.opencypher.tools.tck.values._

import scala.collection.JavaConverters._

object GremlinCypherValueConverter {

  def toExecutionFailed(e: Throwable): ExecutionFailed = {
    val gremlinRemoteException = Iterator
      .iterate(e)(_.getCause)
      .filter(c => c == null || c.isInstanceOf[ResponseException])
      .next()

    gremlinRemoteException match {
      case r: ResponseException =>
        val errors = GremlinErrors.mappings.filterKeys(k => r.getMessage matches k)

        errors.size match {
          case 0 => throw new RuntimeException(s"Unknown error message `${r.getMessage}`", e)
          case 1 => errors.head._2
          case _ => throw new RuntimeException(s"More than 1 errors `${errors.keySet}` match `${r.getMessage}`", e)
        }
      case _ =>
        throw new RuntimeException(
          "Unable to find org.apache.tinkerpop.gremlin.driver.exception.ResponseException in stack ",
          e)
    }
  }

  def toGremlinParams(params: Map[String, CypherValue]): util.Map[String, Object] = {
    new util.HashMap[String, AnyRef](params.mapValues {
      case s: CypherString =>
        s.toString
          .stripPrefix("'")
          .stripSuffix("'")
      case i: CypherInteger => i.value
      case d: CypherFloat   => d.value
      case b: CypherBoolean => b.value
      case n                => throw new IllegalArgumentException(s"Unable to convert param $n")
    }.mapValues(_.asInstanceOf[Object]).asJava)
  }

  def toCypherValueRecords(query: String, results: util.List[util.Map[String, AnyRef]]): CypherValueRecords = {
    val rows = results.asScala
      .map(
        javaMap => javaMap.asScala.mapValues((v) => toCypherValue(v)).toMap
      )
      .toList

    val header =
      if (results.isEmpty) List.empty[String]
      else results.get(0).keySet().asScala.toList

    emptyHeaderWorkaround(query, header, rows)
  }

  private def emptyHeaderWorkaround(query: String, header: List[String], rows: List[Map[String, CypherValue]]) = {
    if (rows.isEmpty) {
      val columns = CypherAst.parse(query, new util.HashMap[String, Any]).statement.returnColumns
      CypherValueRecords.emptyWithHeader(columns)
    } else {
      CypherValueRecords(header, rows)
    }
  }

  private def toCypherValue(v: Any): CypherValue = v match {
    case v: Vertex         => toCypherNode(v)
    case e: Edge           => toCypherRelationship(e)
    case s: String         => CypherString(s)
    case d: lang.Double    => CypherFloat(d)
    case n: Number         => CypherInteger(n.longValue())
    case b: lang.Boolean   => CypherBoolean(b)
    case null              => CypherNull
    case m: util.Map[_, _] => toCypherPropertyMap(m)
    case p: util.List[_]   => toCypherList(p)
  }

  private def toCypherRelationship(e: Edge) = {
    CypherRelationship(e.label(), toCypherPropertyMap(e.properties()))
  }

  private def toCypherNode(v: Vertex) = {
    val labels = if (Vertex.DEFAULT_LABEL.equals(v.label())) Set.empty[String] else Set(v.label())
    CypherNode(labels, toCypherPropertyMap(v.properties()))
  }

  private def toCypherPropertyMap[A <: Property[B], B](iterator: util.Iterator[A]): CypherPropertyMap = {
    val map = iterator.asScala
      .map(v => (v.key(), toCypherValue(v.value())))
      .toMap

    CypherPropertyMap(map)
  }

  private def toCypherPropertyMap(javaMap: util.Map[_, _]): CypherPropertyMap = {
    val map = javaMap.asScala.map { case (k, v) => (k.toString, toCypherValue(v)) }.toMap

    CypherPropertyMap(map)
  }

  private def toCypherList(gremlinList: util.List[_]): CypherValue = {
    toCypherPath(gremlinList) match {
      case Some(path) =>
        path
      case None =>
        val list = gremlinList.asScala
          .map(e => toCypherValue(e))
          .toList

        CypherOrderedList(list)
    }
  }

  private def toCypherPath(gremlinPath: util.List[_]): Option[CypherPath] = {
    val gremlinPathScala = gremlinPath.asScala

    if (gremlinPathScala.nonEmpty && gremlinPathScala.head.isInstanceOf[Vertex]) {
      val startingNode = toCypherNode(gremlinPathScala.head.asInstanceOf[Vertex])
      val vertexes = gremlinPathScala
        .filter(_.isInstanceOf[Vertex])
        .map(_.asInstanceOf[Vertex])
        .map(v => (v.id(), v))
        .toMap[Object, Vertex]

      val relationships = gremlinPathScala
        .grouped(2)
        .filter(it => it.head.isInstanceOf[Vertex] && it.last.isInstanceOf[Edge])
        .map(
          it => {
            val v = it.head.asInstanceOf[Vertex]
            val e = it.last.asInstanceOf[Edge]

            if (e.outVertex().equals(v)) {
              Forward(toCypherRelationship(e), toCypherNode(vertexes(e.inVertex().id())))
            } else {
              Backward(toCypherRelationship(e), toCypherNode(vertexes(e.outVertex().id())))
            }
          }
        )
        .toList

      if (vertexes.nonEmpty) Some(CypherPath(startingNode, relationships)) else None
    } else {
      None
    }
  }
}

File: tinkerpop/cypher-gremlin-neo4j-driver/src/main/java/org/opencypher/gremlin/neo4j/driver/GremlinCypherValueConverter.java
/*
 * Copyright (c) 2018 "Neo4j, Inc." [https://neo4j.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.opencypher.gremlin.neo4j.driver;

import static java.lang.String.format;
import static org.opencypher.gremlin.translation.ReturnProperties.ALL_PROPERTIES;
import static org.opencypher.gremlin.translation.ReturnProperties.ID;
import static org.opencypher.gremlin.translation.ReturnProperties.INV;
import static org.opencypher.gremlin.translation.ReturnProperties.LABEL;
import static org.opencypher.gremlin.translation.ReturnProperties.OUTV;
import static org.opencypher.gremlin.translation.ReturnProperties.isNode;
import static org.opencypher.gremlin.translation.ReturnProperties.isPath;
import static org.opencypher.gremlin.translation.ReturnProperties.isRelationship;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.apache.tinkerpop.gremlin.structure.Vertex;
import org.neo4j.driver.internal.InternalNode;
import org.neo4j.driver.internal.InternalPath;
import org.neo4j.driver.internal.InternalRecord;
import org.neo4j.driver.internal.InternalRelationship;
import org.neo4j.driver.v1.Record;
import org.neo4j.driver.v1.Value;
import org.neo4j.driver.v1.Values;
import org.neo4j.driver.v1.types.Entity;

class GremlinCypherValueConverter {

    static Record toRecord(Map<String, Object> map) {
        List<String> keys = new ArrayList<>();
        List<Value> values = new ArrayList<>();

        for (Map.Entry<String, Object> entry : map.entrySet()) {
            keys.add(entry.getKey());
            values.add(toCypherValue(entry.getValue()));
        }

        return new InternalRecord(keys, values.toArray(new Value[values.size()]));
    }

    private static Value toCypherValue(Object value) {
        if (isNode(value)) {
            return toCypherNode((Map<?, ?>) value).asValue();
        } else if (isRelationship(value)) {
            return toCypherRelationship((Map<?, ?>) value).asValue();
        } else if (isPath(value)) {
            return toCypherPath((List) value);
        } else {
            return Values.value(value);
        }
    }

    @SuppressWarnings("unchecked")
    private static Value toCypherPath(List p) {
        boolean isNode = true;

        Entity[] objects = new Entity[p.size()];
        for (int i = 0; i < p.size(); i++) {
            if (isNode) {
                objects[i] = toCypherNode((Map<?, ?>) p.get(i));
            } else {
                objects[i] = toCypherRelationship((Map<?, ?>) p.get(i));
            }

            isNode = !isNode;
        }

        return new InternalPath(objects).asValue();
    }

    private static InternalRelationship toCypherRelationship(Map<?, ?> e) {
        Long start = toCypherId(e.get(OUTV));
        Long end = toCypherId(e.get(INV));

        Map<String, Value> properties = toCypherPropertyMap(e);

        Object id = e.get(ID);
        Object label = e.get(LABEL);

        return new InternalRelationship(toCypherId(id), start, end, String.valueOf(label), properties);
    }

    private static InternalNode toCypherNode(Map<?, ?> v) {
        Set<String> labels = new HashSet<>();
        String label = String.valueOf(v.get(LABEL));
        if (!Vertex.DEFAULT_LABEL.equals(label)) {
            labels.add(label);
        }

        Map<String, Value> properties = toCypherPropertyMap(v);

        return new InternalNode(toCypherId(v.get(ID)), labels, properties);
    }

    private static Map<String, Value> toCypherPropertyMap(Map<?, ?> e) {
        Map<String, Value> properties = new HashMap<>();
        e.entrySet().stream()
            .filter((n) -> !ALL_PROPERTIES.contains(String.valueOf(n.getKey())))
            .forEach((n) -> properties.put(
                String.valueOf(n.getKey()),
                toCypherValue(n.getValue())));

        return properties;
    }

    private static Long toCypherId(Object id) {
        try {
            return Long.valueOf(String.valueOf(id));
        } catch (Exception e) {
            throw new IllegalArgumentException(format("Entity id should be numeric, got `%s` (%s):", id.getClass(), id));
        }
    }
}
